src/Package/Impl/ProjectSystem/Commands/SourceFilesCommand.cs

```diff
@@ -21,33 +27,70 @@
             if ((commandId == RPackageCommandId.icmdSourceSelectedFiles || commandId == RPackageCommandId.icmdSourceSelectedFilesWithEcho) && nodes.GetSelectedNodesPaths().Count() > 0) {
                 foreach (var n in nodes) {
                     if (n.IsFolder || !Path.GetExtension(n.FilePath).EqualsIgnoreCase(".r")) { 
-                        return CommandStatusResult.Unhandled;
+                        return Task.FromResult(CommandStatusResult.Unhandled);
                     }
                 }
-                return new CommandStatusResult(true, commandText, CommandStatus.Enabled | CommandStatus.Supported);
+                return Task.FromResult(new CommandStatusResult(true, commandText, CommandStatus.Enabled | CommandStatus.Supported));
             }
-            return CommandStatusResult.Unhandled;
+            return Task.FromResult(CommandStatusResult.Unhandled);
         }
 
-        public bool TryHandleCommand(IImmutableSet<IProjectTree> nodes, long commandId, bool focused, long commandExecuteOptions, IntPtr variantArgIn, IntPtr variantArgOut) {
+        public async Task<bool> TryHandleCommandAsync(IImmutableSet<IProjectTree> nodes, long commandId, bool focused, long commandExecuteOptions, IntPtr variantArgIn, IntPtr variantArgOut) {
             if (commandId == RPackageCommandId.icmdSourceSelectedFiles || commandId == RPackageCommandId.icmdSourceSelectedFilesWithEcho) {
-                var rFiles = nodes.GetSelectedNodesPaths().Where(x =>
+                bool echo = commandId == RPackageCommandId.icmdSourceSelectedFilesWithEcho;
+                IEnumerable<string> rFiles = Enumerable.Empty<string>();
+
+                var workflow = _interactiveWorkflowProvider.GetOrCreate();
+                var session = workflow.RSession;
+                if(session.IsHostRunning && session.IsRemote) {
+                    var outputWriter = workflow.ActiveWindow.InteractiveWindow.OutputWriter;
+                    var properties = _configuredProject.Services.ExportProvider.GetExportedValue<ProjectProperties>();
+
+                    string projectDir = Path.GetDirectoryName(_configuredProject.UnconfiguredProject.FullPath);
+                    string projectName = properties.GetProjectName();
+                    string remotePath = await properties.GetRemoteProjectPathAsync();
+
+                    var files = nodes.GetSelectedNodesPaths().Where(x =>
                                Path.GetExtension(x).EqualsIgnoreCase(".r") &&
                                File.Exists(x));
-                bool echo = commandId == RPackageCommandId.icmdSourceSelectedFilesWithEcho;
-                _interactiveWorkflowProvider.GetOrCreate().Operations.SourceFiles(rFiles, echo);
+                    outputWriter.WriteLine("Compressing files...");
+                    IFileSystem fs = new FileSystem();
+                    string compressedFilePath = fs.CompressFiles(files, projectDir, new Progress<string>((p) => {
```

Comment:
It might get frozen. Do you recommend using   _coreShell.DispatchOnUIThread?

---
In reply to: [80771341](https://github.com/Microsoft/RTVS/pull/2405#discussion_r80771341) [](ancestors = 80771341,80536996,80395418)